### PR TITLE
Shrink block device volumes

### PR DESCRIPTION
**Pull request by exarkun**
Fixes https://clusterhq.atlassian.net/browse/FLOC-1503

Initially I thought this would involve improving state discovery to include the size of non-manifest datasets.  I changed my mind, though.  Resizes are only attempted for datasets that have a manifestation and those datasets already have their size discovered.  I left the design notes for non-manifest dataset size discovered in, though, marked `FLOC-1503-x`.  I may go ahead and implement that stuff here anyway, it's barely an extra code and improves the system.  If the reviewer would prefer me not to, though, I'll split those things off into a separate issue.

### CODE REVIEW

File: flocker/node/agents/blockdevice.py [hunk 1]
```diff
@@ -487,6 +487,17 @@ def run(self, deployer):
         )
         unmount = UnmountBlockDevice(dataset_id=self.dataset_id)
         detach = DetachVolume(dataset_id=self.dataset_id)
+
+        # FLOC-1503
+        #
+        # If the volume is shrinking, arrange the changes like
+        #
+        # unmount, detach, resize volume, attach, resize filesystem, mount
```
Comment (itamarst) [inline flocker/node/agents/blockdevice.py]: That seems wrong, don't you need to resize the filesystem first?

**Comment by itamarst** [comment]
Rejecting since the problem I see is either a complete misunderstanding on my part or a fundamental problem.

File: flocker/node/agents/blockdevice.py (hunk 1 above)
Comment (exarkun) [inline flocker/node/agents/blockdevice.py]: Oops.  I reversed the cases.

**Comment by wallrj** [comment]
Thanks @exarkun 

You've just talked me through the proposed changes here and they make sense.

Please proceed and file follow up issues and add links for 
 * Resizing non-manifest datasets
 * Having calculate_changes emit the constituent resize state changes rather than the current compound state change.

File: flocker/control/schema/types.yml [hunk 2]
```diff
@@ -217,8 +217,10 @@ definitions:
     # on what this can be set to for the ZFS backend.  Could probably be
     # considered more carefully.
     minimum: 67108864
-    # This is how you require integers, of course.
-    divisibleBy: 1
+    # Filesystem initialization in the block device backend requires sizes to
+    # be a multiple of 512.
+    multipleOf: 512
```
Comment (itamarst) [inline flocker/control/schema/types.yml]: I suggest making this 1024, insofar as you've just added code that presumes KiB.

File: flocker/node/agents/blockdevice.py [hunk 3]
```diff
@@ -487,19 +513,29 @@ def run(self, deployer):
...
-                ]
-            ),
-            deployer,
-        )
+
+        resize_filesystem = ResizeFilesystem(volume=volume, size=self.size)
+        resize_volume = ResizeVolume(volume=volume, size=self.size)
+        if self.size < volume.size:
+            changes = [
+                unmount,
```
Comment (itamarst) [inline flocker/node/agents/blockdevice.py]: Is there a follow-up issue for what happens if this fails half-way through?

File: flocker/node/agents/test/test_blockdevice.py [hunk 4]
```diff
@@ -2787,65 +2846,148 @@ class ResizeFilesystemInitTests(
...
+                volume=volume, filesystem=filesystem,
+            ),
+            deployer
+        ))
+
+        if new_size > original_size:
+            api.detach_volume(volume.blockdevice_id)
+            api.resize_volume(volume.blockdevice_id, new_size)
+            api.attach_volume(volume.blockdevice_id, host)
+
```
Comment (itamarst) [inline flocker/node/agents/test/test_blockdevice.py]: `ResizeBlockDeviceDataset` detaches before resizing filesystem when shrinking, so this doesn't quite match actual interaction pattern.

**Comment by itamarst** [comment]
Looks good: address, merge forward, merge if buildbot is green.

File: flocker/node/agents/test/test_blockdevice.py (hunk 4 above)
Comment (exarkun) [inline flocker/node/agents/test/test_blockdevice.py]: You can only resize a filesystem when the volume is attached.  In the shrink case, `ResizeBlockDeviceDataset` does the filesystem resize and **then** detaches the volume for the volume resize part of the operation.

**Pull request merged**